zerver/worker/queue_processors.py

```diff
@@ -481,29 +482,36 @@ def consume(self, event: Dict[str, Any]) -> None:
         self.ensure_timer()
 
     def ensure_timer(self) -> None:
-        if self.timer_event is not None:
+        if not self.lock.acquire(blocking=False):
             return
-        self.timer_event = Timer(self.TIMER_FREQUENCY, MissedMessageWorker.maybe_send_batched_emails, [self])
-        self.timer_event.start()
+
+        try:
+            if self.timer_event is not None:
+                return
+
+            self.timer_event = Timer(self.TIMER_FREQUENCY, MissedMessageWorker.maybe_send_batched_emails,
+                                     [self])
+            self.timer_event.start()
+        finally:
+            self.lock.release()
 
     def stop_timer(self) -> None:
-        if self.timer_event and self.timer_event.is_alive():
-            self.timer_event.cancel()
-            self.timer_event = None
+        self.timer_event = None
```

Comment:
I don't understand part of the change -- can you explain why we don't still need to cancel a timer (if running) while holding the lock?

Reply:
I guess maybe part of the point is that we're already holding the lock when this is called, and we know that we're running the timer-handler, so we believe there shouldn't be a timer event going?  

Might be worth some comments explaining this.

Reply:
I think having a one-liner `stop_timer` method doing just that was contibuting to the confusion - I inlined this into `maybe_send_batched_emails` and added a comment - we don't need to cancel a timer because by assumption, there's no timer running when this is executed (unless I screwed up the locking somewhere)